### PR TITLE
Add initial napalm connection plugin

### PR DESCRIPTION
**Pull request by Qalthos**
##### SUMMARY

Adds a connection plugin which uses napalm to make and manage device connections.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

plugins/connection/napalm

##### ANSIBLE VERSION

```paste below
2.8
```

### CODE REVIEW

**Comment by Qalthos** [comment]
cc @abenokraitis

**Comment by abenokraitis** [comment]
cc @thaumos and @samccann

File: lib/ansible/plugins/connection/napalm.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+      - Configures the device platform network operating system.  This value is
+        used to load a napalm device abstraction.
+    vars:
+      - name: ansible_network_os
+  remote_user:
+    description:
+      - The username used to authenticate to the remote device when the SSH
+        connection is first established.  If the remote_user is not specified,
+        the connection will use the username of the logged in user.
+      - Can be configured form the CLI via the C(--user) or C(-u) options
```
Comment (samccann) [inline lib/ansible/plugins/connection/napalm.py]: "Can be configured *from* the ..."

And missing period at  the end.

File: lib/ansible/plugins/connection/napalm.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        key: host_key_auto_add
+    env:
+      - name: ANSIBLE_HOST_KEY_AUTO_ADD
+  persistent_connect_timeout:
+    type: int
+    description:
+      - Configures, in seconds, the amount of time to wait when trying to
+        initially establish a persistent connection.  If this value expires
+        before the connection to the remote device is completed, the connection
+        will fail
```
Comment (samccann) [inline lib/ansible/plugins/connection/napalm.py]: missing period at the end.

File: lib/ansible/plugins/connection/napalm.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        key: connect_timeout
+    env:
+      - name: ANSIBLE_PERSISTENT_CONNECT_TIMEOUT
+  persistent_command_timeout:
+    type: int
+    description:
+      - Configures, in seconds, the amount of time to wait for a command to
+        return from the remote device.  If this timer is exceeded before the
+        command returns, the connection plugin will raise an exception and
+        close
```
Comment (samccann) [inline lib/ansible/plugins/connection/napalm.py]: missing period at the end.

File: lib/ansible/plugins/connection/napalm.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+  password:
+    description:
+      - Configures the user password used to authenticate to the remote device
+        when first establishing the SSH connection.
+    vars:
+      - name: ansible_password
+      - name: ansible_ssh_pass
+  private_key_file:
+    description:
+      - The private SSH key or certificate file used to to authenticate to the
```
Comment (samccann) [inline lib/ansible/plugins/connection/napalm.py]: "used *to* authenticate"
(delete the extra "to"

File: lib/ansible/plugins/connection/napalm.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        key: private_key_file
+    env:
+      - name: ANSIBLE_PRIVATE_KEY_FILE
+    vars:
+      - name: ansible_private_key_file
+  timeout:
+    type: int
+    description:
+      - Sets the connection time, in seconds, for the communicating with the
+        remote device.  This timeout is used as the default timeout value for
```
Comment (samccann) [inline lib/ansible/plugins/connection/napalm.py]: "for communicating with"  (remove the extra 'the')

File: lib/ansible/plugins/connection/napalm.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    description:
+      - Specifies the remote device FQDN or IP address to establish the SSH
+        connection to.
+    default: inventory_hostname
+    vars:
+      - name: ansible_host
+  port:
+    type: int
+    description:
+      - Specifies the port on the remote device to listening for connections
```
Comment (samccann) [inline lib/ansible/plugins/connection/napalm.py]: "remote device *that listens* for"

**Review by samccann — APPROVED** [review]

File: lib/ansible/plugins/connection/__init__.py [hunk 7]
```diff
@@ -360,7 +360,8 @@ def set_implementation_plugin_options(self, task_keys=None, var_options=None, di
         initialize implementation plugin options
         '''
         for plugin in self._implementation_plugins:
-            plugin.set_options(task_keys=task_keys, var_options=var_options, direct=direct)
+            if hasattr(plugin, 'set_options'):
+                plugin.set_options(task_keys=task_keys, var_options=var_options, direct=direct)
```
Comment (bcoca) [inline lib/ansible/plugins/connection/__init__.py:364]: just cause it is 'slightly' more efficient for the 'happy path', put in try/catch and just use set_options, then capture and ignore the attribute error

**Pull request merged**